File: src/main/java/com/portofino/realtrainmodunofficial/client/ClientRenderProfiler.java
package com.portofino.realtrainmodunofficial.client;

import com.portofino.realtrainmodunofficial.RealTrainModUnofficial;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Font;
import net.minecraft.client.gui.GuiGraphics;
import net.neoforged.api.distmarker.Dist;
import net.neoforged.bus.api.SubscribeEvent;
import net.neoforged.fml.common.EventBusSubscriber;
import net.neoforged.neoforge.client.event.RenderGuiEvent;
import org.lwjgl.glfw.GLFW;

/**
 * 軽量なクライアント側描画プロファイラ。
 * 1 秒単位で描画カテゴリごとの合計時間を集計して、必要な時だけ HUD に出す。
 */
@EventBusSubscriber(modid = RealTrainModUnofficial.MODID, value = Dist.CLIENT)
public final class ClientRenderProfiler {
    private static final String[] CATEGORY_NAMES = {"Rail", "Train", "Object"};
    private static final int CATEGORY_RAIL = 0;
    private static final int CATEGORY_TRAIN = 1;
    private static final int CATEGORY_OBJECT = 2;

    private static final long[] totalsNs = new long[CATEGORY_NAMES.length];
    private static final int[] counts = new int[CATEGORY_NAMES.length];
    private static final long[] displayTotalsNs = new long[CATEGORY_NAMES.length];
    private static final int[] displayCounts = new int[CATEGORY_NAMES.length];

    private static long lastSnapshotNs = System.nanoTime();
    private static boolean overlayEnabled;

    private ClientRenderProfiler() {
    }

    public static void toggleOverlay() {
        overlayEnabled = !overlayEnabled;
    }

    public static long begin() {
        return System.nanoTime();
    }

    public static void endRail(long startNs) {
        record(CATEGORY_RAIL, startNs);
    }

    public static void endTrain(long startNs) {
        record(CATEGORY_TRAIN, startNs);
    }

    public static void endInstalledObject(long startNs) {
        record(CATEGORY_OBJECT, startNs);
    }

    private static synchronized void record(int category, long startNs) {
        long elapsed = System.nanoTime() - startNs;
        totalsNs[category] += elapsed;
        counts[category]++;
        snapshotIfNeeded();
    }

    private static void snapshotIfNeeded() {
        long now = System.nanoTime();
        if (now - lastSnapshotNs < 1_000_000_000L) {
            return;
        }
        System.arraycopy(totalsNs, 0, displayTotalsNs, 0, totalsNs.length);
        System.arraycopy(counts, 0, displayCounts, 0, counts.length);
        for (int i = 0; i < totalsNs.length; i++) {
            totalsNs[i] = 0L;
            counts[i] = 0;
        }
        lastSnapshotNs = now;
    }

    @SubscribeEvent
    public static void onRenderGui(RenderGuiEvent.Post event) {
        Minecraft mc = Minecraft.getInstance();
        if (mc.player == null || mc.screen != null || !overlayEnabled) {
            return;
        }

        snapshotIfNeeded();

        GuiGraphics graphics = event.getGuiGraphics();
        Font font = mc.font;
        int x = 8;
        int y = 8;
        int lineHeight = font.lineHeight + 2;
        int width = 0;
        String[] lines = new String[CATEGORY_NAMES.length + 1];
        lines[0] = "Profiler [F8]";
        for (int i = 0; i < CATEGORY_NAMES.length; i++) {
            double totalMs = displayTotalsNs[i] / 1_000_000.0D;
            double avgMs = displayCounts[i] > 0 ? totalMs / displayCounts[i] : 0.0D;
            lines[i + 1] = CATEGORY_NAMES[i] + ": "
                + String.format(java.util.Locale.ROOT, "%.2f ms", totalMs)
                + " / "
                + String.format(java.util.Locale.ROOT, "%.2f avg", avgMs)
                + " (" + displayCounts[i] + ")";
        }

        for (String line : lines) {
            width = Math.max(width, font.width(line));
        }

        graphics.fill(x - 4, y - 4, x + width + 6, y + lineHeight * lines.length + 2, 0x90000000);
        for (int i = 0; i < lines.length; i++) {
            graphics.drawString(font, lines[i], x, y + i * lineHeight, 0xFFFFFF, false);
        }
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/client/ClientRenderProfiler.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.client

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewed
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiGraphicsExtractor
import net.neoforged.api.distmarker.Dist
import net.neoforged.bus.api.SubscribeEvent
import net.neoforged.fml.common.EventBusSubscriber
import net.neoforged.neoforge.client.event.RenderGuiEvent
import java.util.Locale
import kotlin.math.max

@EventBusSubscriber(modid = RealTrainModRenewed.MODID, value = [Dist.CLIENT])
object ClientRenderProfiler {
    private val categoryNames = arrayOf("Rail", "Train", "Object")
    private const val CATEGORY_RAIL = 0
    private const val CATEGORY_TRAIN = 1
    private const val CATEGORY_OBJECT = 2

    private val totalsNs = LongArray(categoryNames.size)
    private val counts = IntArray(categoryNames.size)
    private val displayTotalsNs = LongArray(categoryNames.size)
    private val displayCounts = IntArray(categoryNames.size)

    private var lastSnapshotNs = System.nanoTime()
    private var overlayEnabled = false

    @JvmStatic
    fun toggleOverlay() {
        overlayEnabled = !overlayEnabled
    }

    @JvmStatic
    fun begin(): Long = System.nanoTime()

    @JvmStatic
    fun endRail(startNs: Long) {
        record(CATEGORY_RAIL, startNs)
    }

    @JvmStatic
    fun endTrain(startNs: Long) {
        record(CATEGORY_TRAIN, startNs)
    }

    @JvmStatic
    fun endInstalledObject(startNs: Long) {
        record(CATEGORY_OBJECT, startNs)
    }

    @Synchronized
    private fun record(category: Int, startNs: Long) {
        val elapsed = System.nanoTime() - startNs
        totalsNs[category] += elapsed
        counts[category]++
        snapshotIfNeeded()
    }

    private fun snapshotIfNeeded() {
        val now = System.nanoTime()
        if (now - lastSnapshotNs < 1_000_000_000L) {
            return
        }
        totalsNs.copyInto(displayTotalsNs)
        counts.copyInto(displayCounts)
        totalsNs.fill(0L)
        counts.fill(0)
        lastSnapshotNs = now
    }

    @JvmStatic
    @SubscribeEvent
    fun onRenderGui(event: RenderGuiEvent.Post) {
        val minecraft = Minecraft.getInstance()
        if (minecraft.player == null || minecraft.screen != null || !overlayEnabled) {
            return
        }

        snapshotIfNeeded()

        val graphics: GuiGraphicsExtractor = event.guiGraphics
        val font = minecraft.font
        val x = 8
        val y = 8
        val lineHeight = font.lineHeight + 2
        val lines = Array(categoryNames.size + 1) { index ->
            if (index == 0) {
                "Profiler [F8]"
            } else {
                val category = index - 1
                val totalMs = displayTotalsNs[category] / 1_000_000.0
                val avgMs = if (displayCounts[category] > 0) totalMs / displayCounts[category] else 0.0
                categoryNames[category] + ": " +
                    String.format(Locale.ROOT, "%.2f ms", totalMs) +
                    " / " +
                    String.format(Locale.ROOT, "%.2f avg", avgMs) +
                    " (" + displayCounts[category] + ")"
            }
        }

        var width = 0
        for (line in lines) {
            width = max(width, font.width(line))
        }

        graphics.fill(x - 4, y - 4, x + width + 6, y + lineHeight * lines.size + 2, 0x90000000.toInt())
        for (i in lines.indices) {
            graphics.text(font, lines[i], x, y + i * lineHeight, 0xFFFFFF, false)
        }
    }
}
